sdk/test-utils/recorder-new/MIGRATION.md

```diff
@@ -0,0 +1,160 @@
+
+## Changes to NPM scripts
+
+For the unified recorder client library to work, the [test proxy server] must be active while you are running your tests. Helpers have been added to the `dev-tool` package which manage starting and stopping the test proxy server before and after your tests are run. These helpers use the test proxy server Docker image. Therefore, Docker must be installed in your development environment.
+
+The script defined in `package.json` to run the browser tests should be replaced with `dev-tool run test:browser`, and the script to run the Node tests should look something like `dev tool run test:node-ts-input --mocha=\"--timeout 1200000 'test/**/*.spec.ts'`. In the case of the Node tests, the `--mocha="..."` parameter can be used to specify arguments to be passed through to Mocha. For example:
+
+| Script              | Before migration                                                                                                                                                                         | After migration                                                                                                                        |
+| :------------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------- |
+| `unit-test:browser` | `karma start --single-run`                                                                                                                                                               | `dev-tool run test:browser`                                                                                                            |
+| `unit-test:node`    | `mocha -r esm -r ts-node/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 1200000 --full-trace --exclude \"test/**/browser/*.spec.ts\" \"test/**/*.spec.ts\"` | `dev-tool run test:node-ts-input --mocha=\"--timeout 1200000 --full-trace --exclude 'test/**/browser/*.spec.ts' 'test/**/*.spec.ts'\"` |
+
+## Initializing the recorder
+
+The approach taken to initialize the recorder depends on whether the SDK being tested uses Core v1 (`core-http`) or Core v2 (`core-rest-pipeline`). If your SDK is on Core v2, read on. If you're still on Core v1, jump to the section on Core v1 below.
+
+### For Core v2 SDKs
+
+The recorder is implemented as a custom policy which should be attached to your client's pipeline. Firstly, initialize the recorder:
+
+```ts
+let recorder: TestProxyHttpClient;
+
+/*
+ * Note the use of function() instead of the arrow syntax. We need access to `this` so we
+ * can pass test information from Mocha to the recorder.
+ */
+beforeEach(function(this: Context) {
+  recorder = new TestProxyHttpClient(this.currentTest);
+});
```

Comment:
Could it be possible to avoid `this` references and instead pass the concrete parameters to the recorder?

If not now, in the future?

Reply:
(If you like this idea, this could be a separate issue, to avoid blocking this PR)

Reply:
Can you give an example of what you mean? I'm not sure that it's possible to avoid.

Reply:
Well, the recorder only uses some things from the test context. If they were provided through parameters, it would be more obvious to what’s going on -> easier to debug perhaps? It’s a fundamental change, but for fundamental changes it’s now or never

Reply:
I think I find just being able to pass in a single `this.currentTest` parameter more appealing, to be honest. Otherwise, every package using the recorder will have to pass in the same multiple parameters in their tests. The current approach reduces the duplication to just the one parameter, so is less error prone. I don't know 🤷

Reply:
Depending on how the properties are used, it could be possible to spread `…this` I wonder.

This is a stretch anyway, I don’t require this change, nor will ask for others to agree with my opinion here 😆.